### ISSUE TITLE
Request: Treat generic mapped types with some intersection constraints as homomorphic

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#50563]
# Suggestion

## 🔍 Search Terms

keyof, intersection, mapped type, constraint, homomorphic

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I would like this mapped type to be homomorphic:
```ts
type Remap<T> = {
  [K in keyof T & string]: string
}
```

## 📃 Motivating Example

It's possible to achieve homomorphism for this type with:
```ts
type Remap<T> = {
  [K in keyof T as K & string]: string
}
```

So it's not like my feature request would bring anything new to the table. However, I find the DX of the proposal to be better than this, and perhaps it could be considered.

## 💻 Use Cases

I'm mainly interested in the improved DX because TS already provides the means to achieve this. I will totally understand if this gets just closed as something that won't be done. I couldn't find an issue about this, so I've figured out that it's still worth reporting this even if only for this request to exist in the issue tracker for the people to come.

I've even prototyped~ some of the changes required to make this happen [here](https://github.com/Andarist/TypeScript/commit/121685136badef878c734403512f6423ca9488ac). This code needs cleanup and perhaps it needs to be more strict about those intersections etc, I could refine it further.

[This issue](https://github.com/microsoft/TypeScript/issues/45582) is somewhat related to this one and perhaps it's even an argument why this one here shouldn't be implemented or why objects with index signatures should be exempt from the proposed changes.

**Comment by fatcerberus** [comment]
It's not homomorphic by definition though, because you've (potentially) excluded some keys from consideration *a priori*.  So at that point you're just iterating over a set of strings rather than `T` directly.  Even if the end result of both cases is the same, the way you got there is different - the second case technically keeps all keys and some get dropped only as a side effect of "renaming" them to `never`.

**Comment by Andarist** [comment]
I know how the implementation is different here etc. The difference is rather subtle though - I don't think that most people realize it. It's also a little bit of a bummer that the homomorphic version (what people often want) is more verbose. The mapped type key remapping is a fairly new addition to TypeScript - it's good that we have that option now and that it solves this problem, but I wonder though if it wouldn't be a desirable change if that syntax wouldn't be a thing. And if this would be a desirable thing change 4.1 what makes it non-desirable today? I guess my main problem with this is that things like homomorphism and distributivity rely on such nuance in the syntax that a lot of people just won't spot the difference between those declarations. It feels "implicit".

**Comment by fatcerberus** [comment]
> I guess my main problem with this is that things like homomorphism and distributivity rely on such nuance in the syntax that a lot of people just won't spot the difference between those declarations. It feels "implicit".

Yeah, no argument here.  I have this same gripe with distributive conditional types--not only do they conflate two orthogonal concerns (conditionality and distributivity), but they're only distributive when given a naked type parameter as input, which is... subtle.  I would much prefer dedicated syntax for distribution so that 1) it's more obvious what I'm doing, and 2) I don't have to write "conditional" types that aren't actually conditional.
